### PR TITLE
feat(oas): medusa-oas-cli as OAS build tool

### PR DESCRIPTION
**Pull request by patrick-medusajs**
## What

Introduce a CLI for extracting OAS from the core `medusa` package.

## Why

We need to decouple OAS tooling from documentation tooling in order to allow packages and external systems to leverage our OAS has a dependency.

## How

Introduce a new OAS workspace within packages in order to organize current and future OAS related package. Only 1 OAS package for now.

Introduce a new CLI only package to act as the main gateway for all upcoming OAS tooling. Only 1 command for now.

Update documentation tooling pertaining to OAS to use the CLI instead.

### Prerequisite
From the monorepo root:
* `yarn install`
* `yarn build`

#### Case - validation only - success
* Run `yarn openapi:generate --dry-run`
* Expect console output `🟢 Valid OAS` but no mention of `🔵 Exported OAS`

#### Case - validation only - invalid
* Introduce a bug by renaming `@schema Cart` to `@schema Kart` in [models/cart.ts](https://github.com/medusajs/medusa/blob/0adb0d9ff96087613b7a634cbc97a7e301b01121/packages/medusa/src/models/cart.ts#L2)
* Run `yarn build` to update `@medusajs/medusa` package with the bug.
* Run `yarn openapi:generate --dry-run`
* Expect console output `🔴 Invalid OAS` with a stack trace of the issue.

#### Case - docs generation
* Run `yarn openapi:generate`
* Expect `docs/api/` directory to contain:
  * `admin.oas.json` (raw OAS)
  * `store.oas.json` (raw OAS)
  * `admin.oas.yaml` (sanitized OAS)
  * `store.oas.yaml` (sanitized OAS)
  * `admin/` (updated redocly split output)
  * `store/` (updated redocly split output)

#### Case - crawl additional paths
* From a local medusa server (`medusa-starter-default`), add an `index.ts` file in `src/models/`
* In the `index.ts`, add dummy OAS JSDoc like `/** @schema Foobar */`
* From the root of the monorepo, run `yarn medusa-oas oas --type store --paths path-to-medusa-server/src`
* Expect a `store.oas.json` to be created at the root of the monorepo.
* The `store.oas.json` should contain an additional `Foobar` entry in `components.schemas`.

### CODE REVIEW

File: packages/oas/medusa-oas-cli/README.md
Comment (patrick-medusajs) [inline packages/oas/medusa-oas-cli/README.md:34]:
```suggestion
`yarn medusa-oas oas --type admin`
```
`oas` in missing in all code examples

File: packages/oas/medusa-oas-cli/README.md
Comment (patrick-medusajs) [inline packages/oas/medusa-oas-cli/README.md:38]:
```suggestion
Specify in which directory should the files be outputted. Accepts relative and absolute path. It the directory doesn't exist,
```

File: packages/oas/medusa-oas-cli/README.md
Comment (shahednasser) [inline packages/oas/medusa-oas-cli/README.md:18]:
~~~suggestion
```bash
yarn medusa-oas <command>
```
~~~

File: packages/oas/medusa-oas-cli/README.md
Comment (shahednasser) [inline packages/oas/medusa-oas-cli/README.md:3]:
```suggestion
A command-line tool for all OpenAPI Specifications (OAS) related tooling.
```

File: packages/oas/medusa-oas-cli/README.md
Comment (shahednasser) [inline packages/oas/medusa-oas-cli/README.md:22]:
```suggestion
This command will scan the `@medusajs/medusa` package in order to extract JSDoc OAS into a json file.
```

File: packages/oas/medusa-oas-cli/README.md
Comment (shahednasser) [inline packages/oas/medusa-oas-cli/README.md:24]:
```suggestion
By default, the command will output the two files `admin.oas.json` and `store.oas.json` in the same directory that
```

File: packages/oas/medusa-oas-cli/README.md
Comment (shahednasser) [inline packages/oas/medusa-oas-cli/README.md:25]: Should be on the same line as the previous sentence.

File: packages/oas/medusa-oas-cli/README.md
Comment (shahednasser) [inline packages/oas/medusa-oas-cli/README.md:34]:
~~~suggestion
```
yarn medusa-oas --type admin
```
~~~

File: packages/oas/medusa-oas-cli/README.md
Comment (shahednasser) [inline packages/oas/medusa-oas-cli/README.md:38]:
```suggestion
Specify in which directory should the files be outputted. It accepts a relative or absolute path. If the directory doesn't exist,
```

File: packages/oas/medusa-oas-cli/README.md
Comment (shahednasser) [inline packages/oas/medusa-oas-cli/README.md:41]:
~~~suggestion
```bash
yarm medusa-oas --out-dir
```
~~~

File: packages/oas/medusa-oas-cli/README.md
Comment (shahednasser) [inline packages/oas/medusa-oas-cli/README.md:45]:
```suggestion
Allows passing additional directory paths to crawl for JSDoc OAS and include in the generated OAS.
```

File: packages/oas/medusa-oas-cli/README.md
Comment (shahednasser) [inline packages/oas/medusa-oas-cli/README.md:46]:
```suggestion
It accepts multiple entries. It has to be the last option. Alternatively, ` -- ` can be used before the next option.
```

File: packages/oas/medusa-oas-cli/README.md
Comment (shahednasser) [inline packages/oas/medusa-oas-cli/README.md:48]:
~~~suggestion
```bash
yarn medusa-oas --paths ~/medusa-server/src
```
~~~

File: packages/oas/medusa-oas-cli/README.md
Comment (shahednasser) [inline packages/oas/medusa-oas-cli/README.md:54]:
~~~suggestion
```bash
yarn medusa-oas --dry-run
```
~~~

File: packages/oas/medusa-oas-cli/README.md
Comment (shahednasser) [inline packages/oas/medusa-oas-cli/README.md:58]:
```suggestion
Ignore OAS errors and attempt to output generated OAS files.
```

File: packages/oas/medusa-oas-cli/README.md
Comment (shahednasser) [inline packages/oas/medusa-oas-cli/README.md:60]:
~~~suggestion
```bash
yarn medusa-oas --force
```
~~~

**Review by shahednasser — CHANGES_REQUESTED** [review]
When I ran case 3 to generate documentation files then ran the documentation websites, the API reference pages crashed, but not sure if it's just on my end. I tried it out a couple of times, but still the same result.

I also think we should remove the spec3 files in `docs/api` to avoid confusion since we'll now use the `medusa/oas/*` files.

I left some suggestions for the README, but feel free to accept/ignore those 😄

File: packages/oas/medusa-oas-cli/.npmignore
Comment (adrien2p) [inline packages/oas/medusa-oas-cli/.npmignore:1]: todo: rm in favor or the package.json files: ["dist"]

File: packages/oas/medusa-oas-cli/package.json
Comment (adrien2p) [inline packages/oas/medusa-oas-cli/package.json:6]: todo: `"files": ["dist"]`

File: packages/oas/medusa-oas-cli/src/command-oas.ts
Comment (adrien2p) [inline packages/oas/medusa-oas-cli/src/command-oas.ts:65]: suggestion: move the await isDir to a constant

File: packages/oas/medusa-oas-cli/src/command-oas.ts
Comment (adrien2p) [inline packages/oas/medusa-oas-cli/src/command-oas.ts:73]: suggestion: Same as the previous one

File: packages/oas/medusa-oas-cli/src/command-oas.ts
Comment (adrien2p) [inline packages/oas/medusa-oas-cli/src/command-oas.ts:130]: suggestion: for all, should we use the logger instead of console.log? `import logger from "@medusajs/medusa-cli/dist/reporter"`

**Review by adrien2p — CHANGES_REQUESTED** [review]

File: packages/oas/medusa-oas-cli/src/command-oas.ts
Comment (adrien2p) [reply packages/oas/medusa-oas-cli/src/command-oas.ts:130]: yes you right,it will be better when it will be in its own package. maybe add a todo on that to not forget?

**Comment by patrick-medusajs** [comment]
> When I ran case 3 to generate documentation files then ran the documentation websites, the API reference pages crashed, but not sure if it's just on my end. I tried it out a couple of times, but still the same result.

Ah snap, must be a new circular ref when I updated the PR with `develop`. I'll hunt it down and fix it.

> I also think we should remove the spec3 files in `docs/api` to avoid confusion since we'll now use the `medusa/oas/*` files.

I wasn't sure if this PR should remove them or if we should wait for the release. Makes sense to remove them now.

> I left some suggestions for the README, but feel free to accept/ignore those 😄

Thank you, I'll include them.

cc @shahednasser

File: packages/oas/medusa-oas-cli/src/command-oas.ts
Comment (patrick-medusajs) [reply packages/oas/medusa-oas-cli/src/command-oas.ts]: I actually remove this if statement since the `outDir` will be automatically created using `mkdir` recursive.

File: packages/oas/medusa-oas-cli/src/command-oas.ts
Comment (patrick-medusajs) [reply packages/oas/medusa-oas-cli/src/command-oas.ts]: Slight spin on the suggestion. I created a helper method instead to abstract the complexity of `lstat`. See 9bc39d1b9af81567f9f92b4db72f83633ceb2e70

File: packages/oas/medusa-oas-cli/.npmignore
Comment (patrick-medusajs) [reply packages/oas/medusa-oas-cli/.npmignore]: Ah yes, I much prefer that approach. Done in b05953ed70696f10cb783522fde39362906578d6.

File: packages/oas/medusa-oas-cli/package.json
Comment (patrick-medusajs) [reply packages/oas/medusa-oas-cli/package.json:6]: Done in b05953ed70696f10cb783522fde39362906578d6

File: scripts/build-openapi.js
Comment (patrick-medusajs) [inline scripts/build-openapi.js:75]: @shahednasser I added a new step to the API docs build script that will output circular references that have not been sanitized/patched. That's how I was able to update `docs-utils/redocly/config.yaml` without going completely insane 🤣

File: packages/oas/medusa-oas-cli/README.md
Comment (patrick-medusajs) [reply packages/oas/medusa-oas-cli/README.md]: It is put on a separate line automatically by my IDE since the sentence is longer than 120 characters.

**Review by shahednasser — APPROVED** [review]
LGTM! One thing we should think of though: I can see we have to manually add the circular reference schams into `config.yaml`. So, if at any point in the future someone forgets to add a circular reference it will not be possible to detect it through our current tests and can cause the documentation to crash if we miss it. Do you think there's a way to detect or prevent this?

At the moment the API reference files are generated on merging to `master` and a PR is created, so we can review it before merging to make sure that it's working properly, but again I'm not 100% sure this is the best way to go at it as we can miss it due to human errors.

**Comment by patrick-medusajs** [comment]
> One thing we should think of though: I can see we have to manually add the circular reference schams into config.yaml. So, if at any point in the future someone forgets to add a circular reference it will not be possible to detect it through our current tests and can cause the documentation to crash if we miss it. Do you think there's a way to detect or prevent this?

I could reshuffle `build-openapi.js` to throw an error on circularReferenceCheck failure. My challenge will be to not output files when using `--dry-run` but I think I can make it work using Node's `os.tmpdir` and `fs.mkdtemp` to output into a temporary folder on the host system.

@shahednasser Would you mind if I create a separate PR for the improvement to circularReferenceCheck?

**Comment by patrick-medusajs** [comment]
@adrien2p does it look good to you for a 1st iteration?

**Review by adrien2p — APPROVED** [review]
LGTM, nice iteration 👍 yes @patrick-medusajs I think it is a nice first one

**Pull request merged**
